<a href="https://colab.research.google.com/github/marayyy/Stock-Analysis/blob/Guvi/Analysis_on_Microsoft_Stocks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project on EDA with Python



**Importing Pandas and Plotly Through the below command**

In [1]:
import pandas as pd

In [2]:
import plotly.graph_objects as go
import plotly.express as px

In [3]:
pip install yfinance

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 50 kB 3.1 MB/s 
     |████████████████████████████████| 62 kB 1.1 MB/s 
     |████████████████████████████████| 4.1 MB 39.0 MB/s 
     |████████████████████████████████| 128 kB 69.6 MB/s 
     |████████████████████████████████| 110 kB 58.8 MB/s 
     |████████████████████████████████| 112 kB 49.8 MB/s 
  Attempting uninstall: requests
    Found existing installation: requests 2.23.0
    Uninstalling requests-2.23.0:
      Successfully uninstalled requests-2.23.0
  Attempting uninstall: html5lib
    Found existing installation: html5lib 1.0.1
    Uninstalling html5lib-1.0.1:
      Successfully uninstalled html5lib-1.0.1
  Attempting uninstall: beautifulsoup4
    Found existing installation: beautifulsoup4 4.6.3
    Uninstalling beautifulsoup4-4.6.3:
      Successfully uninstalled beautifulsoup4-4.6.3


In [4]:
import yfinance as yf

**In Python, Date and Time are not data types of their own, but a module named DateTime can be imported to work with the date as well as time. Datetime module comes built into Python, so there is no need to install it externally. The DateTime module provides some functions to get the current date as well as time.**

In [5]:
import datetime
from datetime import date, timedelta

I have imported date class from the datetime module because the object of datetime class can access strftime() method

**timedelta( )** function under datetime library is used for calculating differences in dates and also can be used for date manipulations.

**strftime( )** Converts a datetime object containing current date and time to different string formats; %Y, %m, %d etc are format codes,This method takes one or more format codes as an argument and returns a formatted string based on it

**today( )** method of date class under DateTime module returns a date object which contains the value of Today’s date

In [6]:
today = date.today()
#The date object containing current date and time is stored in 'today' variable

d1 = today.strftime("%Y-%m-%d") 
# %Y	Year with century as a decimal number | %-m	Month as a decimal number | %-d	Day of the month as a decimal number
end_date = d1

d2 = date.today() - timedelta(days=365)
d2 = d2.strftime("%Y-%m-%d")
start_date = d2

In [7]:
data = yf.download('MSFT', 
                      start=start_date, 
                      end=end_date, 
                      progress=False)

data["Date"] = data.index

data = data[["Date", "Open", "High", "Low", 
             "Close", "Adj Close", "Volume"]]

data.reset_index(drop=True, inplace=True)

data.head()

Date        Open        High         Low       Close   Adj Close  \
0 2022-01-03  335.350006  338.000000  329.779999  334.750000  331.642456   
1 2022-01-04  334.829987  335.200012  326.119995  329.010010  325.955780   
2 2022-01-05  325.859985  326.070007  315.980011  316.380005  313.443024   
3 2022-01-06  313.149994  318.700012  311.489990  313.880005  310.966217   
4 2022-01-07  314.149994  316.500000  310.089996  314.040009  311.124695   

     Volume  
0  28865100  
1  32674300  
2  40054300  
3  39646100  
4  32720000

In [8]:
data.columns

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume'], dtype='object')

In [9]:
data.describe()

Open        High         Low       Close   Adj Close  \
count  251.000000  251.000000  251.000000  251.000000  251.000000   
mean   269.105020  272.496095  265.289721  268.917091  267.542417   
std     26.059584   25.973312   25.460905   25.761774   25.013175   
min    217.550003  220.410004  213.429993  214.250000  213.647903   
25%    245.820000  249.430000  243.900002  247.180000  246.423264   
50%    265.679993  268.100006  262.399994  265.899994  264.591125   
75%    289.959991  292.954987  285.464996  289.744995  287.749786   
max    335.350006  338.000000  329.779999  334.750000  331.642456   

             Volume  
count  2.510000e+02  
mean   3.121929e+07  
std    1.148390e+07  
min    9.200800e+06  
25%    2.333430e+07  
50%    2.904390e+07  
75%    3.529260e+07  
max    9.042890e+07

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 251 entries, 0 to 250
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       251 non-null    datetime64[ns]
 1   Open       251 non-null    float64       
 2   High       251 non-null    float64       
 3   Low        251 non-null    float64       
 4   Close      251 non-null    float64       
 5   Adj Close  251 non-null    float64       
 6   Volume     251 non-null    int64         
dtypes: datetime64[ns](1), float64(5), int64(1)
memory usage: 13.9 KB


In [11]:
data.isnull().sum()

Date         0
Open         0
High         0
Low          0
Close        0
Adj Close    0
Volume       0
dtype: int64

In [12]:
data.shape

(251, 7)

In [13]:
data.Date

0     2022-01-03
1     2022-01-04
2     2022-01-05
3     2022-01-06
4     2022-01-07
         ...    
246   2022-12-23
247   2022-12-27
248   2022-12-28
249   2022-12-29
250   2022-12-30
Name: Date, Length: 251, dtype: datetime64[ns]

From the above Dataset,

The table shows open, high, low, and closing prices for each period
The dataset does not have any null values
Timeframe - December 20th 2021 to December 19th 2022
Attributes:

*   Date - Date
*   Open - Opening price in the given timeframe
*   High - Highest price attained in the given timeframe
*   Low - lowest price attained in the given timeframe
*   Close - Closing price in the given timeframe
*   Adj-Close - Stock price after paying off the dividends
*   Volume - trading volume for the selected period

Plotting through a Candlestick Chart as it is widely used to analyse the price movements of stock prices, currencies, and other financial instruments.

In [14]:
figure = go.Figure(data=[go.Candlestick(x=data["Date"],
                                        open=data["Open"], high=data["High"],
                                        low=data["Low"], close=data["Close"])])

figure.update_layout(title = "Microsoft Stock Price Analysis", xaxis_rangeslider_visible=False)

figure.show()

Plotting through a Bar Plot to analyse continuous variables in the dataframe (open, high, low, close, adj-close), discrete variables(volume)

In [15]:
figure = px.bar(data, x = "Date", y= "Close")
figure.show()

Line Plot using RangeSlider allows us to check the price variation date-wise

As we can see here:



*   The closing rate fell drastically from 342.2 in December 2021 to 240.45 in December 2022
*   The lowest closing rate was 214.25 observed in November 3rd 2022.
*   While the highest was in December 27th 2021, 342.2



In [16]:
figure = px.line(data, x='Date', y='Close', 
                 title='Microsoft Stock Market Analysis with Rangeslider')

figure.update_xaxes(rangeslider_visible=True)

figure.show()

Through a Line Plot here, relationship between two variables in consideration (Date, Open) Date and the opening price of the stock can be analyzed.

In [17]:
figure = px.line(data, x='Date', y='Open', 
                 title='Stock Market Analysis with Time Period Selectors')

figure.update_xaxes(
    rangeselector=dict(
        buttons=list([
            dict(count=1, label="1m", step="month", stepmode="backward"),
            dict(count=6, label="6m", step="month", stepmode="backward"),
            dict(count=3, label="3m", step="month", stepmode="backward"),
            dict(count=1, label="1y", step="year", stepmode="backward"),
            dict(step="all")
            ])
        )
    )
figure.show()

# Observations:

343.15 was the highest opening price which fell down to as low as 217.55
Using Scatter Plot here to additionally find the outliers

In [18]:
figure = px.scatter(data, x='Date', y='Close', range_x=['2021-07-12', '2022-07-11'],
                 title="Stock Market Analysis by Hiding Weekend Gaps")

figure.update_xaxes(
    rangebreaks=[
        dict(bounds=["sat", "sun"])]
        )

figure.show()

# Observations

There is a downward trend/ pattern in the above graph which could indicate weak 
to moderate negative correlation between the date and the closing rates

By hiding the weekend gaps, it is observable that the closing rates at the end of the year are high in comparision to the start of the year